File: src/common/web/api/NetworkCommands.ts
import { CommandComposer } from "../core/messaging/composers/CommandComposer";
import { CommandReplyComposer } from "../core/messaging/composers/CommandReplyComposer";
import { MessageBuilder } from "../core/messaging/composers/MessageBuilder";
import { Command } from "../core/messaging/Command";
import { CommandReply } from "../core/messaging/CommandReply";
import { Message } from "../core/messaging/Message";

/**
 * A generic PING command. Requires a ``PingReply`` reply.
 */
@Message.define("command/general/ping")
export class PingCommand extends Command {
    public readonly payload: string = "PING";

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, chain: Message | null = null): CommandComposer<PingCommand> {
        return messageBuilder.buildCommand(PingCommand, {}, chain);
    }
}

/**
 * Reply to the ``PingCommand``.
 */
@Message.define("command/general/ping/reply")
export class PingReply extends CommandReply {
    public readonly payload: string = "PONG";

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, cmd: PingCommand, success: boolean = true, message: string = ""):
        CommandReplyComposer<PingReply> {
        return messageBuilder.buildCommandReply(PingReply, cmd, success, message);
    }
}

File: src/common/py/api/network_commands.py
from ..core.messaging import (
    Command,
    CommandReply,
    Message,
)
from ..core.messaging.composers import (
    MessageBuilder,
    CommandComposer,
    CommandReplyComposer,
)


@Message.define("command/general/ping")
class PingCommand(Command):
    """
    A generic PING command.

    Notes:
        Requires a ``PingReply`` reply.
    """

    payload: str = "PING"

    @staticmethod
    def build(
        message_builder: MessageBuilder, *, chain: Message | None = None
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(PingCommand, chain)


@Message.define("command/general/ping/reply")
class PingReply(CommandReply):
    """
    Reply to ``PingCommand``.
    """

    payload: str = "PONG"

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: PingCommand,
        *,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(PingReply, cmd, success, message)
